### ISSUE TITLE
Possible falsepositive of CS0121 when calling Task.Run

### ISSUE THREAD

**Issue by yaakov-h** [dotnet/roslyn#11575]
**Version Used**:
Microsoft Visual Studio Enterprise 2015
Version 14.0.25123.00 Update 2

**Steps to Reproduce**:
Compile this:

```
using System.Collections.Generic;
using System.Threading.Tasks;

namespace TaskAmbiguity
{
	public class Foo<TEntity>
	{
		public async Task<IEnumerable<TEntity>> GetAllAsync()
		{
			return await Task.Run<IEnumerable<TEntity>>(GetAllSynchronously);
		}

		public IEnumerable<TEntity> GetAllSynchronously()
		{
			return null;
		}
	}
}
```

**Expected Behavior**:

The code compiles.

**Actual Behavior**:

```
1>------ Build started: Project: TaskAmbiguity, Configuration: Debug Any CPU ------
1>C:\Temp\TaskAmbiguity\TaskAmbiguity\Class1.cs(10,22,10,47): error CS0121: The call is ambiguous between the following methods or properties: 'Task.Run<TResult>(Func<TResult>)' and 'Task.Run<TResult>(Func<Task<TResult>>)'
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by gafter** [comment]
I believe this is as currently specified. The call is ambiguous between the following methods or properties: 'Task.Run<TResult>(Func<TResult>)' instantiated with `Task<IEnumerable<TEntity>>` and 'Task.Run<TResult>(Func<Task<TResult>>)' instantiated with `IEnumerable<TEntity>`.

**Comment by gafter** [comment]
@ljw1004 Can you please confirm this behavior?

**Comment by ljw1004** [comment]
Yes it's by design. Here's a more minimal self-contained repro:

```csharp
using System;
using System.Threading.Tasks;

class Program
{
    static int Main()
    {
        f<int>(() => Main()); // fine
        f<int>(Main);         // CS0121 ambiguous between two overloads

        f1<int>(Main);        // fine
        f2<int>(Main);        // CS0407 Main has the wrong return type

        Func<int> d = Main;   // fine
        Action a = Main;      // CS0407 Main has wrong return type

        return 0;
    }

    static void f<T>(Func<T> lambda) { }
    static void f<T>(Func<Task<T>> lambda) { }

    static void f1<T>(Func<T> lambda) { }
    static void f2<T>(Func<Task<T>> lambda) { }

}
```

@yaakov-h you're puzzled at why both overloads of `f` are considered candidates, even though one of them produces a compile-time error CS0407 if you try to use it.

The answer, in a nutshell, is that when you invoke a function `Main()` then it *doesn't use the function's return type* when deciding which overload to use; it solely looks at the arguments+parameters.

Likewise when you convert a method-group `Main` to a delegate, it again doesn't use the function's return type when deciding whether the conversion is allowed; again it solely looks at arguments+parameters.

This gives a surprising and counter-intuitive behavior... `Func<int> d = Main` and `Action a = Main` *are both identity conversions* even though the second one generates a compile-time error afterwards.

Likewise overload resolution sees its two candidates are identity conversions, and hence it's ambiguous between the two, even though one of them generates a compile-time error.


Here is the relevant portion of the spec: https://github.com/ljw1004/csharpspec/blob/gh-pages/conversions.md#method-group-conversions

>  An implicit conversion ([Implicit conversions](conversions.md#implicit-conversions)) exists from a method group ([Expression classifications](expressions.md#expression-classifications)) to a compatible delegate type. Given a delegate type `D` and an expression `E` that is classified as a method group, an implicit conversion exists from `E` to `D` if `E` contains at least one method that is applicable in its normal form ([Applicable function member](expressions.md#applicable-function-member)) to an argument list constructed by use of the parameter types and modifiers of `D`, as described in the following.

>  The compile-time application of a conversion from a method group `E` to a delegate type `D` is described in the following. Note that the existence of an implicit conversion from `E` to `D` does not guarantee that the compile-time application of the conversion will succeed without error.

**Comment by yaakov-h** [comment]
> This gives a surprising and counter-intuitive behavior

I'll agree with that. Can we improve it to make it less surprising and more intuitive?

**Comment by ljw1004** [comment]
I think best would be to start a new thread with a concrete proposal and invoke @ MadsTorgersen.

Would we change it so that it's *no longer* an identity conversion if the return types don't match? (Would that impact type inference for method group conversion?) Or would we just change the rules of overload resolution to take into account return types?

I have a hard time thinking through the back-compat issues in these areas...

**Comment by gafter** [comment]
We have a proposal to "fix" this: https://github.com/dotnet/roslyn/issues/250

I believe it is backwards compatible.
